src/scheduler/Executor.cpp

```diff
@@ -223,19 +237,33 @@ void Executor::threadPoolThread(int threadPoolIdx)
             releaseClaim();
         }
     }
-}
 
-void Executor::shutdownThreadPoolThread(int threadPoolIdx)
-{
-    const auto& logger = faabric::util::getLogger();
-    logger->debug("Shutting down thread pool thread {}:{}", id, threadPoolIdx);
-
-    threadPoolThreads.at(threadPoolIdx) = nullptr;
+    if (selfShutdown) {
+        logger->debug(
+          "Shutting down thread pool thread {}:{}", id, threadPoolIdx);
+
+        // Note - we have to keep a record of dead threads so we can join them
+        // all when the executor shuts down
+        std::shared_ptr<std::thread> thisThread =
+          threadPoolThreads.at(threadPoolIdx);
+        deadThreads.emplace_back(thisThread);
+
+        // Set this thread to nullptr
+        threadPoolThreads.at(threadPoolIdx) = nullptr;
+
+        // See if any threads are still running
+        bool isFinished = true;
+        for (auto t : threadPoolThreads) {
+            if (t != nullptr) {
+                isFinished = false;
+                break;
+            }
+        }
```

Comment:
I tried a few different stdlib options for doing this check, including `find_if`, `find_not_if` etc. but they all came out way more verbose and unreadable than this, so given that this isn't on the critical path performance-wise I'm inclined to leave the simple loop.